#### Задание 1  
Подберите скорость обучения (eta) и количество итераций

In [285]:
import numpy as np

In [286]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [287]:
# Задача про балл на егэ в зависимости от стажа преподавателя
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]

In [288]:
n = X.shape[0]

eta = 0.01 
n_iter = 5000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 200 == 0: #опытным путем выяснила, что лучше понижать скорость обучения каждые 200 итераций
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [2.08 4.27], MSE = 3047.75
Iteration #200: W_new = [37.17493171  5.2427889 ], MSE = 73.76
Iteration #400: W_new = [43.32551177  4.12747604], MSE = 45.41
Iteration #600: W_new = [44.62336601  3.89213018], MSE = 44.06
Iteration #800: W_new = [44.93664858  3.83532121], MSE = 43.98
Iteration #1000: W_new = [45.02207971  3.8198296 ], MSE = 43.97
Iteration #1200: W_new = [45.04810212  3.81511083], MSE = 43.97
Iteration #1400: W_new = [45.0568656   3.81352171], MSE = 43.97
Iteration #1600: W_new = [45.06009831  3.81293551], MSE = 43.97
Iteration #1800: W_new = [45.06139358  3.81270063], MSE = 43.97
Iteration #2000: W_new = [45.06195302  3.81259919], MSE = 43.97
Iteration #2200: W_new = [45.06221167  3.81255228], MSE = 43.97
Iteration #2400: W_new = [45.06233889  3.81252921], MSE = 43.97
Iteration #2600: W_new = [45.06240508  3.81251721], MSE = 43.97
Iteration #2800: W_new = [45.0624

**Вывод:** достигли минимума, в результате которого веса остаются неизменными до нескольких знаков после запятой на протяжении нескольких итераций и MSE = 43.97

**Задача 2*.** В коде избавляемся от итераций по весам, но тут есть ошибка, исправьте ее  

In [289]:
# по примеру задачи с оценкой в зависимости от стажа преподавателя
X, y

(array([[ 1,  1],
        [ 1,  1],
        [ 1,  2],
        [ 1,  5],
        [ 1,  3],
        [ 1,  0],
        [ 1,  5],
        [ 1, 10],
        [ 1,  1],
        [ 1,  2]]),
 [45, 55, 50, 55, 60, 35, 75, 80, 50, 60])

In [290]:
n = X.shape[0]

eta = 0.01 
n_iter = 2500

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
#     for k in range(W.shape[0]):
#         W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    
    #W -= eta * (1/n * 2 * np.dot(X, y_pred - y))
    ## ИЗМЕНЕНИЯ - ДОБАВИЛА ФУНКЦИЮ ТРАНСПОНИРОВАНИЯ К МАТРИЦЕ X
    W -= eta * (1/n * 2 * np.dot(X.T, y_pred - y))
    if i % 200 == 0: #увеличила до 100
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 10        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [2.08 4.27], MSE = 3047.75
Iteration #200: W_new = [38.38986469  5.02247953], MSE = 65.33
Iteration #400: W_new = [43.99463466  4.00614091], MSE = 44.52
Iteration #600: W_new = [44.89160255  3.84348962], MSE = 43.98
Iteration #800: W_new = [45.03515017  3.81745947], MSE = 43.97
Iteration #1000: W_new = [45.05812303  3.8132937 ], MSE = 43.97
Iteration #1200: W_new = [45.06179952  3.81262702], MSE = 43.97
Iteration #1400: W_new = [45.0623879   3.81252033], MSE = 43.97
Iteration #1600: W_new = [45.06248206  3.81250325], MSE = 43.97
Iteration #1800: W_new = [45.06249713  3.81250052], MSE = 43.97
Iteration #2000: W_new = [45.06249954  3.81250008], MSE = 43.97
Iteration #2200: W_new = [45.06249993  3.81250001], MSE = 43.97
Iteration #2400: W_new = [45.06249999  3.8125    ], MSE = 43.97


**Задача 3*.** Вместо того, чтобы задавать количество итераций, задайте другое условие останова алгоритма - когда веса перестают изменяться меньше определенного порога $\epsilon$.

In [291]:
n = X.shape[0]

eta = 0.01 
#max_iter = 1600

weight_dist = np.inf
min_weight_dist = 1e-8

w = np.array([1, 0.5])

iter_num = 0

while weight_dist > min_weight_dist:
    y_pred = np.dot(X, w)
    grad = (2/n * np.dot(X.T, y_pred - y))
    new_weight = w - eta * grad
    weight_dist = np.linalg.norm(new_weight - w, ord = 2)
    err = calc_mse(y, y_pred)
    if iter_num % 100 == 0: 
        print(f'Iteration #{iter_num}: W_new = {w}, MSE = {round(err,2)}')
    
    iter_num +=1
    w = new_weight
    

Iteration #0: W_new = [1.  0.5], MSE = 3047.75
Iteration #100: W_new = [28.22929764  6.86494171], MSE = 177.43
Iteration #200: W_new = [38.32845066  5.03361602], MSE = 65.33
Iteration #300: W_new = [42.36857287  4.30100215], MSE = 47.39
Iteration #400: W_new = [43.98480618  4.00792316], MSE = 44.52
Iteration #500: W_new = [44.63137328  3.89067818], MSE = 44.06
Iteration #600: W_new = [44.89002963  3.84377484], MSE = 43.98
Iteration #700: W_new = [44.99350398  3.82501136], MSE = 43.97
Iteration #800: W_new = [45.03489844  3.81750512], MSE = 43.97
Iteration #900: W_new = [45.05145812  3.81450228], MSE = 43.97
Iteration #1000: W_new = [45.05808274  3.813301  ], MSE = 43.97
Iteration #1100: W_new = [45.0607329   3.81282044], MSE = 43.97
Iteration #1200: W_new = [45.06179308  3.81262819], MSE = 43.97
Iteration #1300: W_new = [45.0622172   3.81255128], MSE = 43.97
Iteration #1400: W_new = [45.06238687  3.81252052], MSE = 43.97
Iteration #1500: W_new = [45.06245474  3.81250821], MSE = 43.97
I